vision/src/main/java/coppercore/vision/VisionLocalizer.java

```diff
@@ -64,6 +66,72 @@ public VisionLocalizer(
         }
     }
 
+    /**
+     * Returns the X angle to the best target, which can be used for simple servoing with vision.
+     *
+     * @param cameraIndex The index of the camera to use.
+     */
+    public Rotation2d getTargetX(int cameraIndex) {
+        return inputs[cameraIndex].latestTargetObservation.tx();
+    }
+
+    /**
+     * calculates the strafing required for drive to be in line with a specific tag + offset
+     *
+     * @param tagId desired tag to align to
+     * @param desiredCameraIndex camera to use for measurements
+     * @param crossTrackOffsetmeters how much to offset horizontal distance by
+     * @return cross track error from tag + offset
+     */
+    public double getCrossTrackDistanceErrorToTag(
+            int tagId, int desiredCameraIndex, double crossTrackOffsetMeters) {
+        // camera not in vision
+        if (desiredCameraIndex >= inputs.length) {
+            return -1;
+        }
+
+        SingleTagObservation tagObserved = inputs[desiredCameraIndex].latestSingleTagObservation;
+
+        // if tag id doesn't match, we assume we don't have that tag in view
+        // therefore, no distance can be observed
+        if (tagObserved.tagId() != tagId) {
+            return -1;
+        }
+
+        // we can subtract the angle of camera if there is one
+        return tagObserved.distance3D() * Math.sin(tagObserved.tx().minus(new Rotation2d()).getRadians())
+                + crossTrackOffsetMeters;
+    }
+
+    /**
+     * calculates the forward / reverse required for drive to be in line with a specific tag +
+     * offset
+     *
+     * @param tagId desired tag to align to
+     * @param desiredCameraIndex camera to use for measurements
+     * @param alongTrackOffsetMeters offset to prevent vision from calculating into tag
+     * @return along track error from tag + offset
+     */
+    public double getAlongTrackDistanceErrorToTag(
+            int tagId, int desiredCameraIndex, double alongTrackOffsetMeters) {
+        // camera not in vision
+        if (desiredCameraIndex >= inputs.length) {
+            return -1;
+        }
+
+        SingleTagObservation tagObserved = inputs[desiredCameraIndex].latestSingleTagObservation;
+
+        // if tag id doesn't match, we assume we don't have that tag in view
+        // therefore, no distance can be observed
+        if (tagObserved.tagId() != tagId) {
+            return -1;
+        }
+
+        // we can subtract the angle of camera if there is one
+        return tagObserved.distance3D() * Math.cos(tagObserved.ty().minus(new Rotation2d()).getRadians())
```

Comment:
What is ty?

Since we're doing cos() here I think we can use the same angle as used above.

My concern is that tx and ty aren't actually angles. In which case we should figure out what they are